Adding the asset library

Omarley7 opened the pull request:
# Adding the asset Library

- The library page now loads assets from an external source and allows users to select their combination.
- It introduces error handling if the username is not a repository found on the endpoint.
- All shared states are now using Redux via `Access` modules.
- Correct use of naming convention in react hooks in the `envUtil` module.
- Introduces graphQL schema in client/src/schema.graphql
- Improved unit testing structure.
- Playwright now uses a time limit of 200ms on non API click events for E2E testing.

Improvements in E2E testing. Now artifacts from failed tests are stored in GitHub Actions enabling developers to analyze screenshots afterward. The build action is now also using the install script (that covers PlayWright installation) instead of just `yarn install`

*The E2E tests needs to use `user1` as test username. Otherwise the tests **may** fail*

----
## In progress
- [x] Developer documentation on how to use Relay
- [x] User documentation on how to configure endpoints
- [x] Playwright configuration for a 200ms time limit on non-api clicks

----
## After review changes
- [x] Do [not ignore](https://github.com/INTO-CPS-Association/DTaaS/pull/78#discussion_r1234254559) playwright-report
- [x] Update path in [jest config](https://github.com/INTO-CPS-Association/DTaaS/pull/78#discussion_r1234322161)
- [x] Remove props from [cartList](https://github.com/INTO-CPS-Association/DTaaS/pull/78#discussion_r1234281509)
- [x] Verify [Private Route and AuthProvider](https://github.com/INTO-CPS-Association/DTaaS/pull/78#discussion_r1234835689) is being nice.
- [x] Change URL in [env files](https://github.com/INTO-CPS-Association/DTaaS/pull/78#discussion_r1234150771)
- [x] [Update](https://github.com/INTO-CPS-Association/DTaaS/pull/78#discussion_r1234368492) `develop` command in package.json 
- [x] Move asset model to [AssetBoard](https://github.com/INTO-CPS-Association/DTaaS/pull/78#discussion_r1234276794)
- [x] Rename [AssetBoard](https://github.com/INTO-CPS-Association/DTaaS/pull/78#discussion_r1234276794) directory
- [x] [equivalency check](https://github.com/INTO-CPS-Association/DTaaS/pull/78#discussion_r1234296646) of graphQL schema
- [x] GitLab group `dtaas` [not working](https://github.com/INTO-CPS-Association/DTaaS/pull/78#discussion_r1234141896)
- [x] [E2E failing](https://github.com/INTO-CPS-Association/DTaaS/pull/78#issuecomment-1597287938) on before each on PR 
- [x] [Artifacts not uploaded](https://github.com/INTO-CPS-Association/DTaaS/pull/78#discussion_r1234361010) on PR
- [x] Dynamic base URL in [playwright tests](https://github.com/INTO-CPS-Association/DTaaS/pull/78#discussion_r1234174157)
- [x] Update [package diagram](https://github.com/INTO-CPS-Association/DTaaS/pull/78#issuecomment-1597421311)
- [ ] ~~[Make test](https://github.com/INTO-CPS-Association/DTaaS/pull/78#discussion_r1234363914) for `RelayEnvironment.ts`~~

Omarley7 commented:
Hi @prasadtalasila, apologies for the delay in this PR, unfortunately, I got a head injury.
I'll add the missing documentation tonight, but I'm curious to see if anything else is missing from this PR.

You mentioned this wouldn't be part of the next release. Would it make more sense to push this to a new branch instead of Distributed Demo?

prasadtalasila commented:
> Hi @prasadtalasila, apologies for the delay in this PR, unfortunately, I got a head injury. I'll add the missing documentation tonight, but I'm curious to see if anything else is missing from this PR.
> 
> You mentioned this wouldn't be part of the next release. Would it make more sense to push this to a new branch instead of Distributed Demo?

@Omarley7 , hope you are well now.

I will need to look at the PR closely. It makes sense to keep this in a separate branch, but let's wait a bit till the authentication is merged into feature/distributed-demo. We are pretty close to merging the authentication branch.

Based on the changes being proposed in this PR, we can take a call after one or two rounds of code review. Take care.

prasadtalasila commented:
you can check the docs by running 

```
cd docs
pip install -r docs/requirements
mkdocs serve
```

prasadtalasila commented:
The [authentication](https://github.com/INTO-CPS-Association/DTaaS/tree/authentication) branch works fine and the files are the same as the `prasadtalasila/dtaas-public/feature/distributed-demo` branch. So it bugs me.

Omarley7 commented:
> you can check the docs by running
> 
> ```
> pip install -r docs/requirements
> mkdocs serve
> ```

Ah, that's cool! Unfortunately I'm getting the following error on all branches
```shell
Defaulting to user installation because normal site-packages is not writeable
ERROR: Could not open requirements file: [Errno 2] No such file or directory: 'docs/requirements'
```

Am I missing something?

prasadtalasila commented:
@KarstenMalle and @Omarley7 , can you check to see if this PR effects the authentication PR? We need to figure out the extent of conflict so that we can proceed further with merge both the PRs in one or two different branches. I really hope to merge both quickly into feature/distributed-demo branch so that we can continue to build on top of the hard work done by both of you.

prasadtalasila commented:
> Ah, that's cool! Unfortunately I'm getting the following error on all branches
> ```shell
> Defaulting to user installation because normal site-packages is not writeable
> ERROR: Could not open requirements file: [Errno 2] No such file or directory: 'docs/requirements'
> ```
> Am I missing something?

My mistake.

```
pip install -r docs/requirements.txt #the file extension was missing before
mkdocs serve
```

KarstenMalle commented:
> > @KarstenMalle and @Omarley7 , can you check to see if this PR effects the authentication PR? We need to figure out the extent of conflict so that we can proceed further with merge both the PRs in one or two different branches. I really hope to merge both quickly into feature/distributed-demo branch so that we can continue to build on top of the hard work done by both of you.
> 
> I only see a single major breaking change so far:
> 
> * [ ]  The username is not stored with the `UserAccess` module in the authentication branch. This is how we retrieve it.
> 
> We've added some minor updates, that will conflict.
> 
> * GitHub Actions now stores E2E error tracing
> * GitHub Actions uses the install script during the build to get PlayWright with frozen dependencies.
> * Renaming variables to follow React convention
> * other minor stuff
> 
> Once the breaking change has been fixed, I recommend merging the Auth first, then we will rebase this PR.

The change would just need to be in `util/Authentication.ts` in the function `getAndSetUsername` to fix this. Instead of using your previous version of that used `dispatch(setUserName(username))` it would need to use `actions.setUser(username)`.

Omarley7 commented:
> The change would just need to be in `util/Authentication.ts` in the function `getAndSetUsername` to fix this. Instead of using your previous version of `dispatch(setUserName(username))` it would need to use `actions.setUser(username)`.

I just double-checked, seems like you don't have any of the updates we made on the previous [PR](https://github.com/Artin13/DTaaS/pull/1). You're missing the UserAccess file completely like we used in [client/src/route/auth/Signin.tsx](https://github.com/Artin13/DTaaS/pull/1/files#diff-bfafc6ca9a7f8d0a1494cc5a6d14388c5a8cf9e5593549295565819ffc07b6f1)

I suggest we make a separate PR to update the upstream with these missing files first, then rebase your PR #72 and finally this PR. Unfortunately, I don't have time to make the separate PR until tomorrow morning.

prasadtalasila commented:
The E2E tests are failing for me with the following log.
```
Running 12 tests using 4 workers                                                                                                                                                       
                                                                                                                                                                                       
  ✘  1 [firefox] › Auth.test.ts:12:3 › Tests on Authentication Flow › Homepage has correct title and signin link (30.0s)                                                               
  ✘  2 [chromium] › Menu.test.ts:13:3 › Menu Links from first page (Layout) › Menu Links are visible (30.0s)                                                                           
  ✘  3 [firefox] › Menu.test.ts:13:3 › Menu Links from first page (Layout) › Menu Links are visible (30.0s)                                                                            
  ✘  4 [chromium] › Auth.test.ts:12:3 › Tests on Authentication Flow › Homepage has correct title and signin link (30.0s)                                                              
  ✘  5 [chromium] › Menu.test.ts:23:3 › Menu Links from first page (Layout) › Menu Links are clickable (30.0s)                                                                         
  ✘  6 [chromium] › Auth.test.ts:21:3 › Tests on Authentication Flow › Fill the signin fields (30.0s)                                                                                  
  ✘  7 [firefox] › Menu.test.ts:23:3 › Menu Links from first page (Layout) › Menu Links are clickable (30.0s)                                                                          
  ✘  8 [firefox] › Auth.test.ts:21:3 › Tests on Authentication Flow › Fill the signin fields (30.0s)                                                                                   
  ✘  9 [chromium] › Auth.test.ts:28:3 › Tests on Authentication Flow › Account Button Contents and Links (30.0s)                                                                       
  ✘  10 [firefox] › Auth.test.ts:28:3 › Tests on Authentication Flow › Account Button Contents and Links (30.0s)                                                                       
  ✘  11 [chromium] › Auth.test.ts:41:3 › Tests on Authentication Flow › Accessing protected routes without authentication (30.0s)                                                      
  ✘  12 [firefox] › Auth.test.ts:41:3 › Tests on Authentication Flow › Accessing protected routes without authentication (30.0s)                                                       
                                                                                                                                                                                       
                                                                                                                                                                                       
  1) [chromium] › Auth.test.ts:12:3 › Tests on Authentication Flow › Homepage has correct title and signin link     


    Test timeout of 30000ms exceeded while running "beforeEach" hook.

       5 |
       6 | test.describe('Tests on Authentication Flow', () => {
    >  7 |   test.beforeEach(async ({ page }) => {
         |        ^
       8 |     await page.goto('./');
       9 |     await page.getByRole('textbox', { name: 'username' }).fill('user1');
      10 |   });

        at /home/prasad/git/prasadtalasila/DTaaS/client/test/e2e/Auth.test.ts:7:8

    Error: locator.fill: Target closed
    =========================== logs ===========================
    waiting for getByRole('textbox', { name: 'username' })
    ============================================================

       7 |   test.beforeEach(async ({ page }) => {
       8 |     await page.goto('./');
    >  9 |     await page.getByRole('textbox', { name: 'username' }).fill('user1');
         |                                                           ^
      10 |   });
      11 |
      12 |   test('Homepage has correct title and signin link', async ({ page }) => {

        at /home/prasad/git/prasadtalasila/DTaaS/client/test/e2e/Auth.test.ts:9:59

    Pending operations:
      - locator.fill at test/e2e/Auth.test.ts:9:59


    attachment #1: trace (application/zip) ─────────────────────────────────────────────────────────
    test-results/Auth-Tests-on-Authentication-Flow-Homepage-has-correct-title-and-signin-link-chromium/trace.zip
    Usage:

        npx playwright show-trace test-results/Auth-Tests-on-Authentication-Flow-Homepage-has-correct-title-and-signin-link-chromium/trace.zip

    ────────────────────────────────────────────────────────────────────────────────────────────────
.....

```

Omarley7 commented:
@DaloonOfDoom am you take a look at the trace files real quick?

prasadtalasila commented on client/script/install.bash:
```diff
@@ -1,3 +1,3 @@
 yarn install --frozen-lockfile
-npx playwright install-deps
\ No newline at end of file
+yarn playwright install --with-deps
```
good change

prasadtalasila commented on client/src/AppProvider.tsx:
```diff
@@ -1,8 +1,11 @@
 import { CssBaseline } from '@mui/material';
 import { ThemeProvider, createTheme, Theme } from '@mui/material/styles';
+import RelayEnvironment from 'RelayEnvironment';
+import { AuthProvider } from 'components/AuthContext';
```
There is already PrivateRoute component in index.tsx. Do we need this here? May be @KarstenMalle can also provide feedback here.

KarstenMalle replied:
This is not using any of our code so I can't really comment on it. 

We use our `AuthProvider` in `AppProvider.tsx` and then the route we want protected in index are wrapped with `PrivateRoute` as so:

```
 <PrivateRoute>
    <Library />
  </PrivateRoute>
```

prasadtalasila replied:
The dependency is from `PrivateRoute -> AuthContext -> Auth`. Is there a need to introduce `AuthContext` into `AppProvider`?

Omarley7 replied:
I believe `PrivateRoute` can only be used within the context of `AuthProvider`. If removed, we get the following error:
```md
# Error
useAuth must be used within an AuthProvider
    at useAuth (http://localhost:3000/static/js/bundle.js:803:11)
    at PrivateRoute (http://localhost:3000/static/js/bundle.js:1244:60)
    at renderWithHooks (http://localhost:3000/static/js/bundle.js:64425:22)
    at mountIndeterminateComponent (http://localhost:3000/static/js/bundle.js:67711:17)
    at beginWork (http://localhost:3000/static/js/bundle.js:69007:20)
    at HTMLUnknownElement.callCallback (http://localhost:3000/static/js/bundle.js:54017:18)
    at Object.invokeGuardedCallbackDev (http://localhost:3000/static/js/bundle.js:54061:20)
    at invokeGuardedCallback (http://localhost:3000/static/js/bundle.js:54118:35)
    at beginWork$1 (http://localhost:3000/static/js/bundle.js:73992:11)
    at performUnitOfWork (http://localhost:3000/static/js/bundle.js:73239:16)
```

Correct me if I'm wrong @KarstenMalle, but the `authProvider` is the global context, which `PrivateRoute` uses? Regardless it follows the same structure as [893e130](https://github.com/prasadtalasila/DTaaS-Public/commit/893e130c0ef44b91b45cd35c335f6c23bb4bfc94) in PR #77, so I suggest moving this to a separate issue instead of this PR @prasadtalasila

KarstenMalle replied:
We might be misunderstanding each other. As can be seen in PR #77 `AuthContext` is not something we are using, and it is old code (over 2 months) as we are still waiting for the authentication PR to be approved. In the new code `AuthProvider` is the global context which informs `PrivateRoute` on if the user is authenticated or not.

Omarley7 replied:
Ahhh thank you for the clarification! I realise we moved it in this PR to keep things streamlined. I reverted the change in commit [3093d76](https://github.com/INTO-CPS-Association/DTaaS/pull/78/commits/3093d762bc000f66f0cc491bc885badbeae19c66) and will resolve this comment.

prasadtalasila reviewed (commented):
More review comments to come...

prasadtalasila commented:
If the [package diagram](https://github.com/INTO-CPS-Association/DTaaS/blob/feature/distributed-demo/docs/developer/client/uml/package-diagram.puml) needs updating and not too much work, please update the same. Otherwise, let it be and I will take care of it later.

prasadtalasila commented on .gitignore:
```diff
@@ -113,6 +113,7 @@ build/
 # Test reports
 coverage/
 playwright-report/
```
why is `client/playwright` not ignored? This is because the playwright output files were / may have been moved from playwright-report to playwright. Can we revert this change without breaking anything. Otherwise, we need to update the .gitignore.

prasadtalasila replied:
It gets created for code on `feature/distributed-demo` branch.

prasadtalasila replied:
The `.gitignore` has been updated in the latest commit on `feature/distributed-demo`. So this is taken care of.

prasadtalasila commented on client/src/page/Menu.tsx:
```diff
@@ -1,35 +1,33 @@
     setAnchorElUser(event.currentTarget);
-  const handleDrawerOpen = () => dispatch(openMenu());
-  const handleDrawerClose = () => dispatch(closeMenu());
+  const handleDrawerOpen = () => menuAction.open();
```
It looks like you are essentially wrapping the dispatch function in menuAction.open(). If you are, the dispatch is a standard method in redux terminology. Has this consistency been maintained at some other place?

Omarley7 replied:
We've made sure to only use dispatch in the `Access` modules, thus keeping Redux "at an arm's length". I'm not sure if this answers your question or if you want us to take further action here?

prasadtalasila replied:
All the redux code is in `store/` I guess. Does any other part of the codebase knows about redux? If not, your statement about keeping Redux "at an arm's length" is valid. Then the changes made here make sense.

I am a bit puzzled about completely wrapping the redux. You can probably keep the reducer, action, dispatch hidden. But how do you keep the subscribe hidden? The react redux components need to be used across the codebase, at least in the AssetBoard / Cart part of the codebase. Is it not?

Omarley7 replied:
No other part of the codebase knows about Redux, so we should be good.

Since we pass the state through the `Access` module as a Redux state, we can simply subscribe to it directly like this:
```typescript
import useCart from 'store/CartAccess';

function ShoppingCart() {
  const { state, actions } = useCart();
}
```

It works perfectly fine for our use case, but there may be other features of Redux needed in the future, that might prove difficult to utilize. But for now, we've encountered no issues.

DaloonOfDoom replied:
The subscription (state?) is handled by the same hook that provides the wrapped actions. 
We made it this way to have a single point of contact between the store and the rest of the application. 
The state of the slice (part of the complete state) is provided by the useAppSelector in the *Access.ts files.
It this instance, the state of the menu is provided by 

const state = useAppSelector((store) => store.menu);

in client/src/store/AppAccess.ts

prasadtalasila commented on client/src/route/library/Library.tsx:
```diff
@@ -1,27 +1,59 @@
+function useLibraryData() {
+  const assetBoardWithSuspense = (
+    pathToAssets: string,
+    privateRepo?: boolean
```
I forgot to ask this basic question. How are the gitlab credentials passed onto the client/ code?

prasadtalasila replied:
As and when the authentication PR is merged, the corresponding headers get added here right?

Omarley7 replied:
Exactly ;)

prasadtalasila commented on client/src/schema.graphql:
```diff
@@ -0,0 +1,70 @@
+type Query {
```
This looks consistent with gitlab graphql types. Has this equivalency check been made?

Omarley7 replied:
Not sure how to make an equivalency check against the GitLab schema. Their schema is way too big for any online [tools ](https://validator.schema.org/#url=https%3A%2F%2Fgitlab.com%2F-%2Fgraphql-explorer) I could find.

I have manually verified it follows the [reference guide](https://docs.gitlab.com/ee/api/graphql/reference/#pageinfo) and found a few discrepancies, which I have corrected.

Apollo offers some form of [composition checking](https://www.apollographql.com/docs/graphos/delivery/schema-checks/#composition-checks). It would take me a while to figure out, I don't think it is worth it right now.

prasadtalasila commented on client/test/e2e/Auth.test.ts:
```diff
@@ -1,40 +1,40 @@
 
-    await page.locator('[aria-label="Open settings"]').click();
-    await page.locator('text=Logout').click();
+    await page.getByLabel('Open settings').click();
```
Both are filtering down to the same element. I see the advantage of readability. Is there any other advantage beyond readability?

Omarley7 replied:
It is also possible to add filtering options to the query. I strongly prefer this, as it's 1-1 with jest queries.
```ts
getByLabel('Open settings', {name: "myDiv"})
```

prasadtalasila reviewed (commented):
I have not looked too closely at the tests. Will do that after the suggested changes have been made. Thank you.

Omarley7 commented on .github/workflows/client.yml:
```diff
@@ -28,19 +28,25 @@ jobs:
+        uses: actions/upload-artifact@v3
+        with:
+          name: E2E-error-trace
+          path: client/test-results
```
It should upload to the [summary of the job](https://docs.github.com/en/actions/using-workflows/storing-workflow-data-as-artifacts). I didn't manage to only run if the previous job failed, hence I'm using `always()`. The advantage is that the trace can be activated with a simple playwright configuration and uploaded even if the job is not failing if the developer is interested.

here's an [example job](https://github.com/Omarley7/DTaaS-Bachelor-new-GUI/actions/runs/5305497501) with artifacts uploaded.

prasadtalasila replied:
Good, I am after the file you have shown in [here](https://github.com/Omarley7/DTaaS-Bachelor-new-GUI/actions/runs/5305497501). This file was not generated for this PR. Is it only generated for commits and not PRs?

Omarley7 replied:
Hmm, that is very strange. I see no reason why it shouldn't produce artifacts for a PR. Do you have a link where I can see the PR job?

prasadtalasila replied:
https://github.com/INTO-CPS-Association/DTaaS/actions/runs/5305452168

Omarley7 replied:
Since the Action has no failing tests, there's no error trace uploaded. I've removed the warning by only attempting to upload if the previous job (tests) fails.

Omarley7 commented on client/jest.config.json:
```diff
@@ -13,7 +13,11 @@
     "src/index.tsx",
     "src/AppProvider.tsx",
-    "src/store/store.ts"
+    "src/store/store.ts",
```
I'll update the path to `src/store/Redux/store.ts` instead

Omarley7 commented on client/jest.config.json:
```diff
@@ -13,7 +13,11 @@
-    "src/store/store.ts"
+    "src/store/store.ts",
+    "src/util/queries/__generated__/",
+    "src/RelayEnvironment.ts",
```
The file was generated during the Relay installation. It only has had minor changes to use Axios instead of fetch, and will later get authentication headers. I assumed it wouldn't make sense to test since it's mostly generated. What do you think? @prasadtalasila

prasadtalasila replied:
As I understand, the generated skeleton is being manually modified, so it makes sense to test it.

Omarley7 commented on client/package.json:
```diff
@@ -16,11 +16,12 @@
     "clean": "script/clean.bash",
-    "format": "prettier --ignore-path ../.gitignore --write \"**/*.{ts,tsx,css,scss}\""
+    "format": "prettier --ignore-path .eslintignore --write \"**/*.{ts,tsx,css,scss}\"",
+    "relay": "relay-compiler"
```
I'll move `yarn relay` above `yarn build`. In my experience, I haven't always been able to run build without updating with `yarn relay` Can't say exactly why.

In addition, if the developer is using `yarn develop` they would need to use this command as well

prasadtalasila replied:
Makes sense. Should the develop be modified to `"develop": "relay-compiler --validate && npx react-scripts start",`

Omarley7 commented on client/src/components/AssetBoard.tsx:
```diff
@@ -0,0 +1,43 @@
+import * as React from 'react';
+import { Grid } from '@mui/material';
+import { Asset } from 'models/Asset';
```
1 - `Asset.ts` is used in `apiUtil.ts`, `CartList.tsx`, and its Redux counterparts. Since it was used in multiple places, we thought it would make sense to store it globally. That said, without `AssetBoard` there would be no need for an Asset model since all other modules are built around `AssetBoard`. Is this the argument for moving `Asset` to `AssetBoard?`

2 - I'll rename the directory, that's a really good point.

prasadtalasila replied:
yes

Omarley7 commented:
@prasadtalasila Thank you for the thorough review. I've added some comments where I could use some clarification or counter-arguments. I've added :eyes: where I'll get right on it and make the changes in the next commit.

prasadtalasila commented:
> @prasadtalasila Thank you for the thorough review. I've added some comments where I could use some clarification or counter-arguments. I've added 👀 where I'll get right on it and make the changes in the next commit.

Thanks for the explanation and appreciate the development effort.

Omarley7 commented on client/src/components/cart/CartList.tsx:
```diff
@@ -0,0 +1,22 @@
+  assets: Asset[];
+}
+
+function CartList(props: OwnProps) {
```
I'm not sure I totally follow what you mean with the props having baggage. But I can reduce the code in one of the following ways:

1. Remove interface and rename:
```ts
function CartList(props: { assetsList: Asset[] }) {
  return (
    <ul>
      {props.assetsList.map((a, i) => (
        <CartItemRender key={i} asset={a}></CartItemRender>
      ))}
    </ul>
  );
}
```
2. Use the `Access` hook both in ShoppingCart.ts and `CartList.ts` + rename
```ts
import useCart from 'store/CartAccess';

function CartList() {
  const { state } = useCart();
  return (
    <ul>
      {state.assets.map((a, i) => (
        <CartItemRender key={i} asset={a}></CartItemRender>
      ))}
    </ul>
  );
}
```

prasadtalasila replied:
The second code looks clean.

Omarley7 commented on client/config/prod.js:
```diff
@@ -9,4 +9,7 @@ window.env = {
   REACT_APP_WORKBENCHLINK_JUPYTERLAB: '/lab',
   REACT_APP_WORKBENCHLINK_JUPYTERNOTEBOOK: '',
+  REACT_APP_BACKEND_URL_GITLAB:
+    'https://worker15.lab.cps.digit.au.dk/api/graphql',
```
I need a functioning GraphQL URL with an existing repository matching and the correct structure for the E2E test to function. I've added a `ci` env file to contain this URL in commit [6262312](https://github.com/Omarley7/DTaaS-Bachelor-new-GUI/commit/6262312ff53f595645f698fed88708ff99ade1be)

Omarley7 commented on client/playwright.config.ts:
```diff
@@ -17,6 +18,7 @@ const config = {
   ],
   use: {
     baseURL: 'http://localhost:4000/dtaas/',
+    trace: 'retain-on-failure',
```
It works perfectly fine, as long as the `baseURL` is ending with `/`.

Also found a way to make this configuration read from `.env` instead of being hardcoded. But this would require a sizable change to the codebase.

Omarley7 commented:
> The E2E tests are failing for me with the following log.
> ....

I assume you are running this locally with a modified environment? Please verify that the `baseURL` in `playwright.config.ts` is ending with `/`. Otherwise, I would need to know what other configurations you are using in order to know what's going on.


At this point, all your comments should be more or less addressed. I've left a few comments open for further feedback from @prasadtalasila and on for @KarstenMalle.

I have not made a test for `RelayEnvironment` yet, as this is pretty new territory and we need to present out thesis the 22nd @ 10am, so I have to move my focus a bit to the presentation part. Hope these changes are enough to get the PR approved.

prasadtalasila commented:
@Omarley7 and @DaloonOfDoom , I am changing the upstream of PR to a dedicated branch. We will merge PR first to asset-cart branch and then check against the feature/distributed-demo. can you please rebase the PR and tell me when it is ready for the review? Thanks.

Omarley7 commented:
@prasadtalasila I've finally gotten it to work after the new authentication update again. The only thing I won't be able to do is to make a test case for `RelayEnvironment`. I cannot find a viable way to test this, it may be a skill issue I'll admit.

Please let me know if there is anything else you want me to update. I have a few stashes on my computer, on of wich is to move `getAndSetUsername()` functionality to UserAccess module instead for better consistency. But I'd rather do this kind of stuff in a new PR in the future.
-

prasadtalasila commented:
When the `dtaas1/<username>` is not found, there is an error.
![Screenshot from 2023-07-11 14-17-08](https://github.com/INTO-CPS-Association/DTaaS/assets/9206466/e6463777-6b7f-467d-9a18-9194dc723bc3)

It is better to just show the common files and show user private files as empty. This thread of code fragment will be useful when user intentionally removes all the files/directories in his private git repository.

prasadtalasila commented:
> When the `dtaas1/<username>` is not found, there is an error.
> ![Screenshot from 2023-07-11 14-17-08](https://github.com/INTO-CPS-Association/DTaaS/assets/9206466/e6463777-6b7f-467d-9a18-9194dc723bc3)
> 
> It is better to just show the common files and show user private files as empty. This thread of code fragment will be useful when user intentionally removes all the files/directories in his private git repository.
> 

In general, if graphql returns a repisitory not found, then please keep the library contents empty. The repository not found error may come in two scenarios.

1. User repository not found
2. Common repository not found
3. Group not found

Keep the relevant library contents page empty for the first two scenarios and throw up a error only on the contents of library page for the third scenario. The DTaaS application can still continue to function when user goes to "digital twins" or "workbench" pages.

prasadtalasila commented:
@Omarley7 , the improvements made in the latest updates are good. Can you please check the latest comments? Thanks.

Omarley7 commented:
> In general, if graphql returns a repisitory not found, then please keep the library contents empty. The repository not found error may come in two scenarios.
> 
> 1. User repository not found
> 2. Common repository not found
> 3. Group not found
> 
> Keep the relevant library contents page empty for the first two scenarios and throw up a error only on the contents of library page for the third scenario. The DTaaS application can still continue to function when user goes to "digital twins" or "workbench" pages.

I see it may be a bit too aggressive to throw errors on missing group or username. For technical purposes, I've changed the error message in commit [4113619](https://github.com/INTO-CPS-Association/DTaaS/pull/78/commits/411361947de0a6dbeb2d6d63ec95cc2319887c3d) to be displayed in the 'private' or 'common' tabs, if there's an issue fetching the data.

Alternatively, I would need to modify the logic to follow this flow, and somehow add the `emit error` stage. If we choose this approach the error message from the API utility would be lost, unless we use `console.error` or `console.warn`.
```mermaid
graph TB
  LibraryPage --> CategoryTabComponent
  CategoryTabComponent -->|Functions| FunctionTabComponent
  FunctionTabComponent -->|Private| AssetBoard_Private
  FunctionTabComponent -->|Common| AssetBoard_Common
  AssetBoard_Private --> useAssets
  AssetBoard_Common --> useAssets
  useAssets -- Error --> ErrorHandling
  ErrorHandling -.-> |Emit error|FunctionTabComponent
```

I'll take a look at the other comments if I have time this week, otherwise It'll be in august I'll pick this PR back up. I'm currently on vacation in the US.

prasadtalasila commented:
Losing the error message from gitlab graphql service is ok for now. There will be a logger put in later to keep track of this issue. I suggest to hide the error message and show an empty content page.

Your plan for contribution makes sense. Wish you a great vacation.

Message ID: ***@***.***>

prasadtalasila commented:
This PR is fantastic. It brings in all the components to complete the **Library** page. This is being kept in `asset-cart` branch for now. The code from this shall be used on the **Digital Twins** page first and then this will be added to the **Library** page.

The consolidated list of potential improvements in this PR are:

1. Even though the `.github/workflows/client.yml` tries to store the build artifacts, they do not get stored.
2. Some files `utils/` dependent on other code in `src/`. This is not a good convention. It is probably better to move these dependent files of `utils/` to top-level at `src/`.
3. If a user repository is not found in gitlab, the graphql returns a repisitory not found. The front end exits the application completely with a stacktrace. Instead, the front-end has to show an empty library. A proper way to handle this problem is:

```mermaid
graph TB
  LibraryPage --> CategoryTabComponent
  CategoryTabComponent -->|Functions| FunctionTabComponent
  FunctionTabComponent -->|Private| AssetBoard_Private
  FunctionTabComponent -->|Common| AssetBoard_Common
  AssetBoard_Private --> useAssets
  AssetBoard_Common --> useAssets
  useAssets -- Error --> ErrorHandling
  ErrorHandling -.-> |Emit error|FunctionTabComponent
```

Thanks @Omarley7  and @DaloonOfDoom for this good contribution.
